### ISSUE TITLE
[4.0] Patch updates is malfunction for 4.x series, `/media` folder is outdated

### ISSUE THREAD

**Issue by Fedik** [joomla/joomla-cms#35920]
Patch updates is malfunction for 4.x series, and `/media` folder is outdated.
Everyone who installed patch update since 4.0.0 have outdated `/media` folder.

Temporary work around: install `4.0.x-Stable-Update_Package` .

The issue comes from a reason that we have moved source of `/media` folder to `/build/media_source/` and so build script does not see difference. 
https://github.com/joomla/joomla-cms/blob/03dd7a146168de70f5dfc32395ed6f07c0e84232/build/build.php#L430

We need another solution for `/media` diff. 
Maybe need to compare content of `/media` itself, somehow. And not relay on GIT history.

**Comment by laoneo** [comment]
Or just include media also in patch updates.

**Comment by Fedik** [comment]
We have a problem. Everyone who made patch update since 4.0.x to 4.0.4 will not be able to update to next release.
Because after #35388 merged they did not received a JS files that comes with that PR (and old JS files was removed).
On next update they will get JS error.

~I set it as release blocker, we need a solution before 4.0.5~
Ping @wilsonge

**Comment by richard67** [comment]
This will have happened only if they used upload and update and the smaller packages but not when using live update, which always fetches the big update package for update from any version, as far as I know.

**Comment by Fedik** [comment]
@richard67 thanks for info :wink:

**Comment by zero-24** [comment]
Should still be a RB so so make sure our releases are not broken.

**Comment by richard67** [comment]
Yes, just not as urgent as it seemed, but it needs to be fixed of course.

**Comment by wilsonge** [comment]
For now I've decided to not ship the patch packages with Joomla 4.0.5 as this isn't going to be a trivial problem. There's a new flag in the build script to handle that.

**Comment by richard67** [comment]
Meanwhile I've found a bit time to check how our build script works (because I need that knowledge for another thing I'm working on).

For creating the "small" update packages, it uses "git diff" on tags. This would only work if we would have all the build results checked in into our sources (which would be silly).

So we have to compare folders and not tags.

But the build script also deletes stuff in the temporary build folder (with name = some number, located in build/tmp) after having created the full package and before creating the update package.

So we can't just compare the content of that temporary folder with the same of another version (and if we could it would still mean we have to run a build for every version which we compare.

We have the same problem when we want to generate the list of deleted files and folders.

The solution up to now for the latter was to compare folders into which a full installation package has been unpacked before.

The same could be done in the build script, too. It could download and unpack the full packages and then use `git diff --name-status <folder1> <folder2>` to compare these folders in the same way as it now compares tags, but with the difference that it now includes all results the build script, too.

I will try to find the time to work on such a solution.

**Comment by PhilETaylor** [comment]
@zero-24 added the Release Blocker label on 20 Nov 2021

Im guessing, as there have been releases since then, that this is not a release blocker after all and can have that label removed.

**Comment by zero-24** [comment]
@PhilETaylor This is correct after that release an "workaround" has been implemented by not delivering said broken update packages. Still this is an issue that should be looked at and fixed, will take down the label for now while it is defenitly not a ideal solution at all.
